### ISSUE TITLE
[indent] multiple false positives in recent versions

### ISSUE THREAD

**Issue by 43081j** [typescript-eslint/typescript-eslint#121]
**Example**

```JSON
{
  "rules": {
    "@typescript-eslint/indent": "error"
  }
}
```

```TS
export type Foo = A
  | B
  | C;
```

**Expected Result**

Identation is linted.

**Actual Result**

Indentation is not linted.

**Additional Info**

It seems we simply don't deal with type aliases.

---

We should probably be handling `TSIntersectionType` and `TSUnionType`. I am happy to PR this but wasn't sure what the best solution is... maybe to transform into a `LogicalExpression`?

**Comment by 43081j** [comment]
In actual fact, the indent rule seems broken in a few places since the last change to it...

```ts
  private static _store: Map<string, Foo> =
    new Map<string, Foo>();

// linter complains that it doesn't look like this
  private static _store: Map<string, Foo> =
  new Map<string, Foo>();
```

Another:

```ts
  private _endpoint: string =
    '/foo';

// linter complains that it should be
  private _endpoint: string =
  '/foo';
```

Another:

```ts
  protected _messageMap =
    new Map<number, Message>([
      [-1, msg]
    ]);

// linter complains that it should be
  protected _messageMap =
  new Map<number, Message>([
    [-1, msg]
  ]);
```

There are too many instances of this for it to be a coincidence. My guess is something broke the rule in the most recent changes.

This is all with config:

```json
    "@typescript-eslint/indent": ["warn", 2, {
      "FunctionDeclaration": {
        "parameters": 2
      },
      "FunctionExpression": {
        "parameters": 2
      },
      "CallExpression": {
        "arguments": 1
      },
      "SwitchCase": 1
    }],
```

Before I updated to the scoped package from the monorepo, there were no lint errors with the same config and sources.

**Comment by bradzacher** [comment]
confirmed against master.
Need to add more cases for these nodes.
Probably need to transform `TSUnionType`/`TSIntersectionType` to a binary expression or similar.

**Comment by ivanvoznyakovsky** [comment]
this one https://github.com/bradzacher/eslint-plugin-typescript/issues/271 is still an issue in latest release

**Comment by Philipp91** [comment]
`'@typescript-eslint/indent': ['error', 4, {SwitchCase: 1, /* ... Ignore JSX nodes ...*/}]`

Let me add another test case, where currently there's a false positive error:
```
const doSomething = <P extends object, V>(
    {something, ...props}: SomePropsType<P>
): P => ({
    ...props as P, // Here it wants 8 instead of 4 spaces.
    ...something, // Here it wants 8 instead of 4 spaces.
});
```

I simplified it to this, which still throws an error:
```
const test = <T extends object>(
) =>
    true; // Here it wants 8 instead of 4 spaces.
```

Also, removing the generic type parameter and/or the multi-line "parameter list" solves the problem:
```
const test = (
) =>
    true; // Now it agrees with my 4 spaces.
const test = <T extends object>() =>
    true; // Now it agrees with my 4 spaces.
const test = () =>
    true; // Now it agrees with my 4 spaces.
```

**Comment by DacioRomero** [comment]
I'm getting this

```typescript
const foo = [1, 2, 3]

let bar

bar = foo
  .reduce<number>((acc, val) => {
    return acc + val
  }, 0)
```
to throw errors

```
  3:5   warning  'bar' is assigned a value but never used      @typescript-eslint/no-unused-vars
  6:19  warning  Missing return type on function               @typescript-eslint/explicit-function-return-type
  7:1   error    Expected indentation of 2 spaces but found 4  @typescript-eslint/indent
  8:1   error    Expected indentation of 0 spaces but found 2  @typescript-eslint/indent```

**Comment by lostfields** [comment]
I got the same issue with indenting using generics, both of this is valid using indent of 4;

```typescript
let list = [1, 2, 3]
    .map(value => {
        return String(value)
    })

let list = [1, 2, 3]
    .map<string>(value => {
    return String(value)
})
```

**Comment by nolazybits** [comment]
I also have an issue with the indent on enums

```typescript
export enum AnEnum
{
    SOMETHING = "something",
    ELSE = "else",
}
```
with the following rules
```
                "@typescript-eslint/brace-style": ["error", "allman"],
                "@typescript-eslint/indent": [ // https://eslint.org/docs/rules/indent#enforce-consistent-indentation-indent
                    "error",
                    4,
                    {
                        "ArrayExpression": "first",
                        "ObjectExpression": "first",
                        "FunctionDeclaration": {
                            "parameters": "first"
                        },
                        "FunctionExpression": {
                            "parameters": "first"
                        }
                    }
                ],
```

will show an error on the brace on the new line, saying the indentation of 0 should be 4 :/

**Comment by stevenhair** [comment]
I also see this issue with annotations:

```typescript
@injectable()
class FooClass {
    public constructor(
        @inject(mongoClientProviderToken) mongoClientProvider: MongoClientProvider,
        @inject(FooConfig) config: FooConfig
    ) {
        // code
    }
```

eslint wants it to look like:

```typescript
@injectable()
class FooClass {
    public constructor(
    @inject(mongoClientProviderToken) mongoClientProvider: MongoClientProvider,
        @inject(FooConfig) config: FooConfig
    ) {
        // code
    }
```

If I remove the first annotation, everything is fine.

**Comment by AnthonyIacono** [comment]
@stevenhair Same issue here. Did you ever manage to figure out a workaround short of disabling the indentation rule?

**Comment by stevenhair** [comment]
@AnthonyIacono unfortunately not, I just disabled the rule for the file.

**Comment by bradzacher** [comment]
Merging into #1824
